packages/ra-core/src/sideEffect/saga/cancelSaga.js

```diff
@@ -0,0 +1,119 @@
+import { all, call, take, takeEvery, put, race } from 'redux-saga/effects';
+import { delay } from 'redux-saga';
+import { push } from 'react-router-redux';
+
+import {
+    CRUD_UPDATE,
+    CRUD_UPDATE_MANY,
+    CRUD_DELETE,
+    CRUD_DELETE_MANY,
+} from '../../actions/dataActions';
+import {
+    showNotification,
+    hideNotification,
+} from '../../actions/notificationActions';
+import {
+    CANCELLABLE,
+    CANCEL,
+    startOptimisticMode,
+    stopOptimisticMode,
+} from '../../actions/cancelActions';
+import { refreshView } from '../../actions/uiActions';
+import resolveRedirectTo from '../../util/resolveRedirectTo';
+import { setListSelectedIds } from '../../actions/listActions';
+
+function* handleCancelRace(cancellableAction) {
+    const { payload: { action, delay: cancelDelay } } = cancellableAction;
+    // display cancellable notification
+    switch (action.type) {
+        case CRUD_UPDATE: {
+            yield put(
+                showNotification('ra.notification.updated', 'info', {
+                    messageArgs: {
+                        smart_count: 1,
+                    },
+                    cancellable: true,
+                })
+            );
+            break;
+        }
+        case CRUD_UPDATE_MANY: {
+            const actions = [
+                put(
+                    showNotification('ra.notification.updated', 'info', {
+                        messageArgs: {
+                            smart_count: action.payload.ids.length,
+                        },
+                        cancellable: true,
+                    })
+                ),
+            ];
+            if (action.payload.unselectAll) {
+                actions.push(put(setListSelectedIds(action.meta.resource, [])));
+            }
+            yield all(actions);
+            break;
+        }
+        case CRUD_DELETE: {
+            yield put(
+                showNotification('ra.notification.deleted', 'info', {
+                    messageArgs: {
+                        smart_count: 1,
+                    },
+                    cancellable: true,
+                })
+            );
+            break;
+        }
+        case CRUD_DELETE_MANY: {
+            const actions = [
+                put(
+                    showNotification('ra.notification.deleted', 'info', {
+                        messageArgs: {
+                            smart_count: action.payload.ids.length,
+                        },
+                        cancellable: true,
+                    })
+                ),
+            ];
+            if (action.payload.unselectAll) {
+                actions.push(put(setListSelectedIds(action.meta.resource, [])));
+            }
+            yield all(actions);
+            break;
+        }
+    }
+    yield put(startOptimisticMode());
+    // dispatch action in optimistic mode (no fetch)
+    yield put(action);
```

Comment:
I realize that instead of dispatching the same action twice (which has the effect to update the store twice), I'd better dispatch the action with an `_OPTIMISTIC` suffix. That's the type of action that the `ids` and `data` reducers should watch (not the ones dispatching the fetch call). 

That would probably ease the implementation of the Confirm alternatives